File: 496-next-greater-element-i/496-next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Stack<Integer> stk = new Stack<>();
        HashMap<Integer,Integer> map = new HashMap<>();
        int [] ans = new int[nums1.length];
        
        for(int j = 0; j < nums2.length; ++j)
        {
                while (!stk.empty() && stk.peek() < nums2[j])
                {
                    map.put(stk.pop(),nums2[j]);
                }
                stk.push(nums2[j]);
        }
        
       for (int i = 0; i < nums1.length; ++i)
       {
           ans[i] = map.getOrDefault(nums1[i],-1);
       }
        return ans;
    }
}

File: 496-next-greater-element-i/496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        stk = []
        ans = [-1 for i in range(len(nums1))]
        dict = defaultdict(int)
        for i in range(len(nums2)):
            while stk and stk[-1] < nums2[i]:
                top = stk.pop()
                dict[top] = nums2[i]
            stk.append(nums2[i])
        
        for i in range(len(nums1)):
            if nums1[i] in dict:
                ans[i] = dict[nums1[i]]
        
        return ans